core/commonMain/src/kotlinx/serialization/descriptors/SerialDescriptors.kt

```diff
@@ -121,7 +121,25 @@ public fun SerialDescriptor(serialName: String, original: SerialDescriptor): Ser
     return WrappedSerialDescriptor(serialName, original)
 }
 
-internal class WrappedSerialDescriptor(override val serialName: String, original: SerialDescriptor) : SerialDescriptor by original
+internal class WrappedSerialDescriptor(override val serialName: String, private val original: SerialDescriptor) :
+    SerialDescriptor by original {
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other !is WrappedSerialDescriptor) return false
+
+        return serialName == other.serialName && original == other.original
```

Comment:
Is it ok if we check `serialName` inside `original`?

Reply:
Interesting point. Ideally, we should not check it. However, given that `equalsImpl` requires lambda to check type parameters, and we do not have any way to get them here, it seems that we can't avoid this check. 

At least, this should solve the case when the descriptor is identical, but object is created twice.